### PR TITLE
Vaccine Visualization

### PR DESCRIPTION
**Pull request by mnneveau**
Add visualization functionality to vaccine design, outputs vaccine as jpg in same directory as original output file is written

### CODE REVIEW

**Comment by chrisamiller** [comment]
Nice! Can you also update the tests such that they cover the visualization creation as well? (You usually can't do byte-for-byte comparisons of pdfs, because they sometimes contain the date and stuff, but you can check for presence of the file, non-zero size, etc)

**Comment by malachig** [comment]
Cool.  Can we see an example of one of these visualization PDFs?  Maybe post a screenshot here?

**Comment by chrisamiller** [comment]
https://github.com/mnneveau/vaccinevisual/blob/master/vaccine.pdf

**Comment by mnneveau** [comment]
I also put an example into the test_data folder for vaccine_design

**Comment by chrisamiller** [comment]
Still working out some issues with graphics and headless test layers too - hold off on review until those get ironed out, please!

**Comment by malachig** [comment]
This looks nice!

**Comment by jhundal** [comment]
Very nice! Do the numbers represent the junction binding score?  If so, it'd be good to add nM to them.
Another suggestion would be bring the candidate gene names closer to the actual segment. They seem a bit far off.

**Comment by mnneveau** [comment]
I believe this addition is ready for review. The output file has been changed to a jpg, and an example has been added to the test_data dir, or can be viewed here: https://raw.githubusercontent.com/mnneveau/pVAC-Seq/0ef17fd562207ec81f08047b6d5f5336a6864c33/tests/test_data/vaccine_design/vaccine.jpg

File: pvacseq/lib/vaccine_design.py [hunk 1]
```diff
@@ -158,6 +159,221 @@ def anneal(self):
...
+    os.system('convert -density 300 -quality 100 ' + ps_file + " " + out_file)
+    os.system('rm ' + ps_file)
+    return()
+
+def output_vaccine_png(input_file, out_f):
+    min_pep_length = 8
+    max_pep_length = 100
+
+    pep_seqs = []
+    with open(input_file, 'r') as input_f:
```
Comment (susannasiebert) [inline pvacseq/lib/vaccine_design.py]: I would make the parsing of the input file into its own method that returns `pep_seq` and `junct_scores`.

File: pvacseq/lib/vaccine_design.py [hunk 2]
```diff
@@ -158,6 +159,221 @@ def anneal(self):
...
+
+    #Error if not a peptide sequence for every peptide ID
+    if len(pep_ids) != len(pep_seqs):
+        print("Error: Not an equal number of peptide sequences and peptide IDs")
+        sys.exit()
+
+    conversion_factor = get_conversion_factor(pep_seqs)
+
+    #determine number of peptides (not including junction additions)
+    num_peptides = 0
```
Comment (susannasiebert) [inline pvacseq/lib/vaccine_design.py]: This paragraph could also be its own method.

File: pvacseq/lib/vaccine_design.py [hunk 3]
```diff
@@ -158,6 +159,221 @@ def anneal(self):
...
+    #add white space in circle before genes
+    draw_wht_space(t, circle_radius, wht_space_angle)
+    angle_parsed += wht_space_angle
+    draw_junction(t, pen_thin)
+
+    #draw main part of circle
+    junctions_parsed = 0
+    peptides_parsed = 0
+    for pep in pep_seqs:
+        pep_length = len(pep)
```
Comment (susannasiebert) [inline pvacseq/lib/vaccine_design.py]: The content of this for-loop body could also be pulled into a `draw_peptide` subroutine.

File: pvacseq/lib/vaccine_design.py [hunk 4]
```diff
@@ -383,6 +599,10 @@ def main(args_input=sys.argv[1:]):
     if not args.keep_tmp:
         shutil.rmtree(tmp_dir)
 
+    out_f = "/".join((base_output_dir, runname))
```
Comment (susannasiebert) [inline pvacseq/lib/vaccine_design.py]: Using `os.path.join` is recommended over concatenating your own path manually.

**Comment by susannasiebert** [comment]
Before Meghan left, she merged in her PR 324 into this PR as well. I will undo that part and reset this PR to commit 8e5128d so that I can merge in just the visualization part.

**Pull request merged**